File: rent-zone/Services/AuthService.swift
import Foundation

// MARK: - Auth DTOs
struct LoginRequest: Encodable {
    let email: String
    let password: String
}

struct RegisterRequest: Encodable {
    let name: String
    let email: String
    let password: String
    let location: String
}

struct AuthResponse: Decodable {
    let user: UserDTO
    let accessToken: String
    let refreshToken: String
}

struct RefreshResponse: Decodable {
    let accessToken: String
    let refreshToken: String
}

// MARK: - Auth Service
class AuthService {
    static let shared = AuthService()
    private init() {}

    // MARK: Login
    func login(email: String, password: String) async throws -> UserDTO {
        let body: [String: Any] = ["email": email, "password": password]
        let result: AuthResponse = try await APIClient.shared.request(
            endpoint: "/auth/login",
            method: "POST",
            body: body
        )
        TokenStorage.accessToken = result.accessToken
        TokenStorage.refreshToken = result.refreshToken
        TokenStorage.userId = result.user.id
        return result.user
    }

    // MARK: Register
    func register(name: String, email: String, password: String, location: String, university: String? = nil, phoneNumber: String? = nil, preferredCategory: String? = nil) async throws -> UserDTO {
        let body: [String: Any] = [
            "name": name,
            "email": email,
            "password": password,
            "location": location,
            "university": university as Any,
            "phoneNumber": phoneNumber as Any,
            "preferredCategory": preferredCategory as Any
        ]
        let result: AuthResponse = try await APIClient.shared.request(
            endpoint: "/auth/register",
            method: "POST",
            body: body
        )
        TokenStorage.accessToken = result.accessToken
        TokenStorage.refreshToken = result.refreshToken
        TokenStorage.userId = result.user.id
        return result.user
    }

    // MARK: Refresh Tokens
    func refreshTokens() async throws {
        guard let refreshToken = TokenStorage.refreshToken else {
            throw APIError.unauthorized
        }
        let body: [String: Any] = ["refreshToken": refreshToken]
        let result: RefreshResponse = try await APIClient.shared.request(
            endpoint: "/auth/refresh",
            method: "POST",
            body: body
        )
        TokenStorage.accessToken = result.accessToken
        TokenStorage.refreshToken = result.refreshToken
    }

    // MARK: Logout
    func logout() async throws {
        try await APIClient.shared.request(
            endpoint: "/auth/logout",
            method: "POST",
            body: nil,
            authenticated: true
        ) as EmptyResponse
        TokenStorage.clear()
    }

    // MARK: Get Current User
    func getCurrentUser() async throws -> UserDTO {
        return try await APIClient.shared.request(
            endpoint: "/auth/me",
            authenticated: true
        )
    }
}

// MARK: - Empty Response helper
struct EmptyResponse: Decodable {}

File: rent-zone-backend/src/modules/auth/auth.service.ts
import bcrypt from 'bcryptjs';
import { OAuth2Client } from 'google-auth-library';
import appleSignin from 'apple-signin-auth';
import { AccountProvider, CategoryType } from '@prisma/client';
import prisma from '../../config/db';
import {
  signAccessToken,
  signRefreshToken,
  verifyRefreshToken,
  JwtPayload,
} from '../../utils/jwt.utils';
import { sendOtpEmail } from '../../utils/mailer.utils';

// ── Register ──────────────────────────────────────────────────────────────────
export const registerUser = async (
  name: string,
  email: string,
  password: string,
  location: string,
  university?: string,
  phoneNumber?: string,
  preferredCategory?: CategoryType
) => {
  const existing = await prisma.account.findUnique({ where: { email } });
  if (existing) throw new Error('Email already in use');

  const passwordHash = await bcrypt.hash(password, 10);

  const user = await prisma.user.create({
    data: {
      name,
      location,
      university,
      phoneNumber,
      preferredCategory,
      account: { create: { provider: 'email', email, passwordHash } },
    },
    include: { account: { select: { provider: true, email: true } } },
  });

  const payload: JwtPayload = { userId: user.id, email };
  const accessToken = signAccessToken(payload);
  const refreshToken = signRefreshToken(payload);

  await prisma.account.update({ where: { userId: user.id }, data: { refreshToken } });

  const { account, ...userWithoutAccount } = user;
  return { user: { ...userWithoutAccount, provider: account?.provider, email: account?.email }, accessToken, refreshToken };
};

// ── Login ─────────────────────────────────────────────────────────────────────
export const loginUser = async (email: string, password: string) => {
  const account = await prisma.account.findUnique({
    where: { email },
    include: { user: true },
  });

  if (!account || !account.passwordHash) throw new Error('Invalid credentials');

  const isValid = await bcrypt.compare(password, account.passwordHash);
  if (!isValid) throw new Error('Invalid credentials');

  const payload: JwtPayload = { userId: account.userId, email };
  const accessToken = signAccessToken(payload);
  const refreshToken = signRefreshToken(payload);

  await prisma.account.update({ where: { id: account.id }, data: { refreshToken } });

  return { user: account.user, accessToken, refreshToken };
};


const googleClient = new OAuth2Client(process.env.GOOGLE_CLIENT_ID);

async function verifyGoogleToken(idToken: string) {
  const ticket = await googleClient.verifyIdToken({
    idToken,
    audience: process.env.GOOGLE_CLIENT_ID,
  });
  return ticket.getPayload();
}

async function verifyAppleToken(identityToken: string) {
  return appleSignin.verifyIdToken(identityToken, {
    audience: process.env.APPLE_BUNDLE_ID,
  });
}

// ── OAuth Login ───────────────────────────────────────────────────────────────
export const oauthLogin = async (
  provider: AccountProvider,
  idToken: string,
  name?: string, // Fallback if not in token
  location = ''
) => {
  let email: string | undefined;
  let verifiedName: string | undefined = name;

  if (provider === 'google') {
    const payload = await verifyGoogleToken(idToken);
    email = payload?.email;
    verifiedName = payload?.name || name;
  } else if (provider === 'apple') {
    const payload = await verifyAppleToken(idToken);
    email = payload.email;
    // Apple only sends name on first login, so we might need the fallback
  }

  if (!email) throw new Error('Invalid token: email not found');

  let isNewUser = false;
  let account = await prisma.account.findUnique({
    where: { email },
    include: { user: true },
  });

  if (!account) {
    isNewUser = true;
    await prisma.user.create({
      data: {
        name: verifiedName || 'User',
        location,
        account: { create: { provider, email } },
      },
    });
    account = await prisma.account.findUnique({
      where: { email },
      include: { user: true },
    });
  }

  if (!account) throw new Error('Failed to create or find account');

  const payload: JwtPayload = { userId: account.userId, email };
  const accessToken = signAccessToken(payload);
  const refreshToken = signRefreshToken(payload);

  await prisma.account.update({ where: { email }, data: { refreshToken } });

  return { user: (account as any).user, accessToken, refreshToken, isNewUser };
};

// ── Refresh Tokens ────────────────────────────────────────────────────────────
export const refreshTokens = async (token: string) => {
  const payload = verifyRefreshToken(token);

  const account = await prisma.account.findFirst({
    where: { userId: payload.userId, refreshToken: token },
  });
  if (!account) throw new Error('Invalid refresh token');

  const newPayload: JwtPayload = { userId: payload.userId, email: payload.email };
  const accessToken = signAccessToken(newPayload);
  const refreshToken = signRefreshToken(newPayload);

  await prisma.account.update({ where: { id: account.id }, data: { refreshToken } });

  return { accessToken, refreshToken };
};

// ── Logout ────────────────────────────────────────────────────────────────────
export const logoutUser = async (userId: string) => {
  await prisma.account.update({ where: { userId }, data: { refreshToken: null } });
};

// ── Get Current User ──────────────────────────────────────────────────────────
export const getCurrentUser = async (userId: string) => {
  return prisma.user.findUnique({
    where: { id: userId },
    include: { account: { select: { provider: true, email: true } } },
  });
};

export const updateProfile = async (
  userId: string,
  data: {
    name?: string;
    location?: string;
    university?: string;
    phoneNumber?: string;
    profileImage?: string;
    preferredCategory?: CategoryType;
  }
) => {
  return prisma.user.update({
    where: { id: userId },
    data,
    include: { account: { select: { provider: true, email: true } } },
  });
};
// ── OTP Logic ───────────────────────────────────────────────────────────────
export const sendOtp = async (email: string) => {
  const otp = Math.floor(100000 + Math.random() * 900000).toString();
  const expiresAt = new Date(Date.now() + 10 * 60 * 1000); // 10 minutes

  await prisma.oTP.upsert({
    where: { email },
    update: { code: otp, expiresAt },
    create: { email, code: otp, expiresAt },
  });

  await sendOtpEmail(email, otp);
};

export const verifyOtp = async (email: string, code: string) => {
  const otpEntry = await prisma.oTP.findUnique({ where: { email } });

  if (!otpEntry || otpEntry.code !== code) {
    throw new Error('Invalid verification code');
  }

  if (new Date() > otpEntry.expiresAt) {
    throw new Error('Verification code has expired');
  }

  // Delete the OTP after successful verification
  await prisma.oTP.delete({ where: { email } });

  // Check if user already exists
  const account = await prisma.account.findUnique({
    where: { email },
    include: { user: true },
  });

  if (account) {
    const payload: JwtPayload = { userId: account.userId, email };
    const accessToken = signAccessToken(payload);
    const refreshToken = signRefreshToken(payload);
    await prisma.account.update({ where: { id: account.id }, data: { refreshToken } });
    return { user: account.user, accessToken, refreshToken, isNewUser: false };
  }

  return { email, isNewUser: true };
};
